Signal/src/ViewControllers/PhotoLibrary/PhotoLibrary.swift

```diff
@@ -0,0 +1,288 @@
+                }
+
+                guard let dataUTI = dataUTI else {
+                    resolver.reject(PhotoLibraryError.assertionError(description: "dataUTI was unexpectedly nil"))
+                    return
+                }
+
+                guard let dataSource = DataSourceValue.dataSource(with: imageData, utiType: dataUTI) else {
+                    resolver.reject(PhotoLibraryError.assertionError(description: "dataSource was unexpectedly nil"))
+                    return
+                }
+
+                resolver.fulfill((dataSource: dataSource, dataUTI: dataUTI))
+            }
+        }
+    }
+
+    private func requestVideoDataSource(for asset: PHAsset) -> Promise<(dataSource: DataSource, dataUTI: String)> {
+        return Promise { resolver in
+
+            _ = imageManager.requestExportSession(forVideo: asset, options: nil, exportPreset: AVAssetExportPresetMediumQuality) { exportSession, _ in
+
+                guard let exportSession = exportSession else {
+                    resolver.reject(PhotoLibraryError.assertionError(description: "exportSession was unexpectedly nil"))
+                    return
+                }
+
+                exportSession.outputFileType = AVFileType.mp4
+                exportSession.metadataItemFilter = AVMetadataItemFilter.forSharing()
+
+                let exportPath = OWSFileSystem.temporaryFilePath(withFileExtension: "mp4")
+                let exportURL = URL(fileURLWithPath: exportPath)
+                exportSession.outputURL = exportURL
+
+                Logger.debug("starting video export")
+                exportSession.exportAsynchronously {
+                    Logger.debug("Completed video export")
+
+                    guard let dataSource = DataSourcePath.dataSource(with: exportURL, shouldDeleteOnDeallocation: true) else {
+                        resolver.reject(PhotoLibraryError.assertionError(description: "Failed to build data source for exported video URL"))
+                        return
+                    }
+
+                    resolver.fulfill((dataSource: dataSource, dataUTI: kUTTypeMPEG4 as String))
+                }
+            }
+        }
+    }
+
+    func outgoingAttachment(for asset: PHAsset) -> Promise<SignalAttachment> {
+        switch asset.mediaType {
+        case .image:
+            return requestImageDataSource(for: asset).map { (dataSource: DataSource, dataUTI: String) in
+                return SignalAttachment.attachment(dataSource: dataSource, dataUTI: dataUTI, imageQuality: .medium)
+            }
+        case .video:
+            return requestVideoDataSource(for: asset).map { (dataSource: DataSource, dataUTI: String) in
+                return SignalAttachment.attachment(dataSource: dataSource, dataUTI: dataUTI)
+            }
+        default:
+            return Promise(error: PhotoLibraryError.unsupportedMediaType)
+        }
+    }
+}
+
+class PhotoCollection {
+    private let collection: PHAssetCollection
```

Comment:
Just enough info to render the collection in the collection picker without maintaining a `PHFetchResult` for its assets.